File: python/packages/fd-evals/src/fd_evals/claim_grounding.py
"""Claim-grounding-rate reliability metric — per VeriGraph (arXiv:2606.16603).

``claim_grounding_rate = (claims reachable from a source node) / (total claims)``

For a completed run, a *claim* is a sentence in the final agent output and a
*source node* is a tool-step output (raw data the agent observed). Per
VeriGraph's claim-level definition, a claim is **grounded** when a reachable
evidence path exists from some source node to that claim.

This module mirrors the Rust ``fd_otel::claim_grounding`` shape **and
algorithm** token-for-token, so the same run scores identically on either
plane (a shared golden fixture pins the agreement).

Honest scope: "reachable evidence path" is operationalized as a
**deterministic lexical-overlap reachability proxy** — pure, CI-stable, no LLM
judge — in the same spirit as ``fd_evals.firing_rate``. It is *grounding rate
per VeriGraph*, a lineage to the claim-level auditability literature, **not** a
ferrumdeck-original metric and **not** a semantic-entailment judgment. It is a
reliability *signal*: an optional project threshold only *flags* a run, never
blocks a tool or kills a run.
"""

from __future__ import annotations

import json
import re
from collections.abc import Iterable
from dataclasses import dataclass
from typing import Any

from fd_runtime.tracing import FD_RELIABILITY_CLAIM_GROUNDING_DEFAULT_THRESHOLD

# Pinned constants — keep in sync with the Rust `fd_otel::claim_grounding`.
DEFAULT_MIN_CLAIM_GROUNDING_RATE = FD_RELIABILITY_CLAIM_GROUNDING_DEFAULT_THRESHOLD
MIN_CLAIM_TOKENS = 3
GROUNDING_OVERLAP = 0.5
_MIN_TOKEN_LEN = 3

_CLAIM_SPLIT = re.compile(r"[.!?\n]")
_TOKEN_SPLIT = re.compile(r"[^A-Za-z0-9]+")
_STEP_TYPE_TOOL = "tool"


@dataclass(frozen=True)
class ClaimGrounding:
    """Per-run claim-grounding snapshot.

    Same field shape as the Rust ``fd_otel::claim_grounding::ClaimGrounding``
    struct so the JSON wire contract is one shape, not two.
    """

    claims_total: int
    claims_grounded: int
    rate: float
    below_threshold: bool
    threshold: float

    def to_dict(self) -> dict[str, Any]:
        return {
            "claims_total": self.claims_total,
            "claims_grounded": self.claims_grounded,
            "rate": self.rate,
            "below_threshold": self.below_threshold,
            "threshold": self.threshold,
        }

    @property
    def is_empty(self) -> bool:
        return self.claims_total == 0


def compute(
    claims_total: int,
    claims_grounded: int,
    threshold: float = DEFAULT_MIN_CLAIM_GROUNDING_RATE,
) -> ClaimGrounding:
    """Pure compute. Mirrors Rust ``ClaimGrounding::compute``.

    ``claims_grounded`` is clamped to ``claims_total``. A run with no claims
    returns ``rate == 1.0`` and ``below_threshold == False`` — nothing to
    ground means nothing ungrounded, so "no claims" never flags.
    """
    if claims_total < 0 or claims_grounded < 0:
        raise ValueError("claim counters must be non-negative")
    claims_grounded = min(claims_grounded, claims_total)
    rate = (claims_grounded / claims_total) if claims_total > 0 else 1.0
    below = claims_total > 0 and rate < threshold
    return ClaimGrounding(
        claims_total=claims_total,
        claims_grounded=claims_grounded,
        rate=rate,
        below_threshold=below,
        threshold=threshold,
    )


def compute_from_texts(
    output: str,
    sources: Iterable[str],
    threshold: float = DEFAULT_MIN_CLAIM_GROUNDING_RATE,
) -> ClaimGrounding:
    """Derive the metric from the final ``output`` text and ``sources``
    (tool-output strings). Deterministic; mirrors the Rust
    ``compute_from_texts`` rule exactly."""
    source_tokens: set[str] = set()
    for src in sources:
        source_tokens.update(_significant_tokens(src))

    total = 0
    grounded = 0
    for claim in _split_claims(output):
        toks = set(_significant_tokens(claim))
        if len(toks) < MIN_CLAIM_TOKENS:
            continue
        total += 1
        covered = sum(1 for t in toks if t in source_tokens)
        if covered / len(toks) >= GROUNDING_OVERLAP:
            grounded += 1
    return compute(total, grounded, threshold)


def compute_from_run(
    output: Any,
    steps: Iterable[dict[str, Any]],
    threshold: float = DEFAULT_MIN_CLAIM_GROUNDING_RATE,
) -> ClaimGrounding:
    """Derive the metric from a run's final ``output`` and its ``steps``.

    Source nodes are the outputs of ``step_type == "tool"`` steps (raw data the
    agent observed). ``output`` and tool outputs may be strings or structured
    JSON — both are stringified before tokenizing (tokenization strips all
    punctuation, so the token set is identical regardless of JSON formatting,
    which is what keeps the two planes in agreement on structured output).
    """
    output_text = _stringify(output)
    sources: list[str] = []
    for step in steps:
        # Defensive: callers may hand a malformed / non-list `steps` (e.g. a
        # mock run context). Skip anything that isn't a step dict rather than
        # raise — this is a best-effort reliability signal, never fatal.
        if not isinstance(step, dict):
            continue
        if step.get("step_type") == _STEP_TYPE_TOOL:
            out = step.get("output")
            if out is not None:
                sources.append(_stringify(out))
    return compute_from_texts(output_text, sources, threshold)


def record_on_span(span: Any, metric: ClaimGrounding) -> None:
    """Tag an OTel span with the claim-grounding attributes (mirror of the
    Rust ``record_on_span``)."""
    from fd_runtime.tracing import (
        FD_RELIABILITY_CLAIM_GROUNDING_FLAGGED,
        FD_RELIABILITY_CLAIM_GROUNDING_RATE,
        FD_RELIABILITY_CLAIM_GROUNDING_THRESHOLD,
    )

    span.set_attribute(FD_RELIABILITY_CLAIM_GROUNDING_RATE, metric.rate)
    span.set_attribute(FD_RELIABILITY_CLAIM_GROUNDING_FLAGGED, metric.below_threshold)
    span.set_attribute(FD_RELIABILITY_CLAIM_GROUNDING_THRESHOLD, metric.threshold)


def _split_claims(text: str) -> list[str]:
    return [c.strip() for c in _CLAIM_SPLIT.split(text) if c.strip()]


def _significant_tokens(text: str) -> list[str]:
    return [t.lower() for t in _TOKEN_SPLIT.split(text) if len(t) >= _MIN_TOKEN_LEN]


def _stringify(value: Any) -> str:
    if value is None:
        return ""
    if isinstance(value, str):
        return value
    return json.dumps(value, sort_keys=True, ensure_ascii=False)


__all__ = [
    "ClaimGrounding",
    "DEFAULT_MIN_CLAIM_GROUNDING_RATE",
    "GROUNDING_OVERLAP",
    "MIN_CLAIM_TOKENS",
    "compute",
    "compute_from_run",
    "compute_from_texts",
    "record_on_span",
]

File: rust/crates/fd-otel/src/claim_grounding.rs
//! Claim-grounding-rate reliability metric — claim-level grounding per
//! **VeriGraph** (arXiv:2606.16603).
//!
//! `claim_grounding_rate = (claims reachable from a source node) / (total claims)`
//!
//! For a completed run, a *claim* is a sentence in the final agent output and a
//! *source node* is a tool-step output (raw data the agent observed). Per
//! VeriGraph's claim-level definition, a claim is **grounded** when a reachable
//! evidence path exists from some source node to that claim.
//!
//! ## Honest scope — a deterministic proxy, not an LLM judge
//!
//! We operationalize "reachable evidence path" as a **deterministic
//! lexical-overlap reachability proxy**: a claim is grounded when a sufficient
//! fraction of its significant tokens are covered by the union of source-node
//! tokens. This is pure and CI-stable — same inputs → same output on every
//! machine, no model call, no clock — in the same spirit as
//! [`crate::firing_rate`]. It is **"grounding rate per VeriGraph"**, a
//! lineage to the claim-level auditability literature, **not** a
//! ferrumdeck-original metric and **not** a semantic-entailment judgment.
//!
//! ## Reliability signal, not a gate
//!
//! The metric is a *reliability signal*. A project may set an optional
//! `min_claim_grounding_rate` threshold that **flags** a run below it
//! ([`ClaimGrounding::below_threshold`]); it never blocks a tool and never
//! kills a run — the deny-by-default posture is for tool permissions, not for
//! reliability scoring.
//!
//! Anti-pivot guard: like [`crate::firing_rate`], the compute takes plain
//! counters / text (no `fd-storage` dependency); the gateway extracts the
//! output + tool-output strings and calls in.

use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use tracing::Span;

use crate::genai::attrs;

/// Default opt-in minimum grounding rate. Only used when a project explicitly
/// sets `min_claim_grounding_rate`; absent a project threshold the metric is
/// computed and surfaced but **never flags** (off by default).
pub const DEFAULT_MIN_CLAIM_GROUNDING_RATE: f64 = 0.70;

/// A claim must carry at least this many distinct significant tokens to be
/// scored — drops trivial fragments ("Done.", "OK!") that aren't real claims.
pub const MIN_CLAIM_TOKENS: usize = 3;

/// Fraction of a claim's significant tokens that must be covered by a source
/// node for the claim to count as grounded.
pub const GROUNDING_OVERLAP: f64 = 0.5;

/// Minimum length (in ASCII chars) of a "significant" token. Shorter tokens
/// (`a`, `is`, `to`, `of`) are dropped as non-discriminative.
const MIN_TOKEN_LEN: usize = 3;

/// Per-run claim-grounding snapshot. Pure derived metric — same inputs always
/// produce the same struct. Mirrored field-for-field by the Python
/// `fd_evals.claim_grounding.ClaimGrounding`.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct ClaimGrounding {
    /// Total scored claims found in the final output.
    pub claims_total: u32,
    /// Claims with a reachable evidence path from a source node.
    pub claims_grounded: u32,
    /// `claims_grounded / claims_total`, in `[0.0, 1.0]`. `1.0` when
    /// `claims_total == 0` — a run that makes no claims has nothing ungrounded,
    /// so "no claims" is never a failure (mirrors the firing-rate empty window).
    pub rate: f64,
    /// `true` when `claims_total > 0 && rate < threshold`. A *flag* only — never
    /// an enforcement action.
    pub below_threshold: bool,
    /// Threshold used for the flag decision.
    pub threshold: f64,
}

impl ClaimGrounding {
    /// Compute from raw counters. `claims_grounded` is clamped to
    /// `claims_total` so a mis-count can't produce a rate > 1.0.
    pub fn compute(claims_total: u32, claims_grounded: u32, threshold: f64) -> Self {
        let claims_grounded = claims_grounded.min(claims_total);
        let rate = if claims_total == 0 {
            1.0
        } else {
            f64::from(claims_grounded) / f64::from(claims_total)
        };
        let below_threshold = claims_total > 0 && rate < threshold;
        Self {
            claims_total,
            claims_grounded,
            rate,
            below_threshold,
            threshold,
        }
    }

    /// Derive the metric from the final `output` text and the `sources`
    /// (tool-output strings). Deterministic — see the module docs for the
    /// exact lexical-overlap reachability rule. Mirrors the Python
    /// `compute_from_run` token-for-token.
    pub fn compute_from_texts(output: &str, sources: &[String], threshold: f64) -> Self {
        let source_tokens: HashSet<String> =
            sources.iter().flat_map(|s| significant_tokens(s)).collect();

        let mut total = 0u32;
        let mut grounded = 0u32;
        for claim in split_claims(output) {
            let toks = distinct_significant_tokens(&claim);
            if toks.len() < MIN_CLAIM_TOKENS {
                continue;
            }
            total += 1;
            let covered = toks.iter().filter(|t| source_tokens.contains(*t)).count();
            let frac = covered as f64 / toks.len() as f64;
            if frac >= GROUNDING_OVERLAP {
                grounded += 1;
            }
        }
        Self::compute(total, grounded, threshold)
    }

    /// Whether the run made no scored claims (renders as "no claims" rather
    /// than a 100% reading on the dashboard).
    pub fn is_empty(&self) -> bool {
        self.claims_total == 0
    }
}

/// Split text into claim fragments on sentence terminators + newlines.
fn split_claims(text: &str) -> Vec<String> {
    text.split(['.', '!', '?', '\n'])
        .map(|s| s.trim().to_string())
        .filter(|s| !s.is_empty())
        .collect()
}

/// Significant tokens of a text: ASCII-alphanumeric runs of length
/// `>= MIN_TOKEN_LEN`, lowercased. Non-ASCII chars act as separators (a
/// deterministic, locale-free rule that both planes implement identically).
fn significant_tokens(text: &str) -> Vec<String> {
    text.split(|c: char| !c.is_ascii_alphanumeric())
        .filter(|t| t.len() >= MIN_TOKEN_LEN)
        .map(|t| t.to_ascii_lowercase())
        .collect()
}

/// Distinct (deduplicated, insertion-order-independent) significant tokens.
fn distinct_significant_tokens(text: &str) -> HashSet<String> {
    significant_tokens(text).into_iter().collect()
}

/// Tag a [`tracing::Span`] with the claim-grounding attributes under the
/// existing `ferrumdeck.*` semconv extension.
pub fn record_on_span(span: &Span, metric: &ClaimGrounding) {
    span.record(
        attrs::FERRUMDECK_RELIABILITY_CLAIM_GROUNDING_RATE,
        metric.rate,
    );
    span.record(
        attrs::FERRUMDECK_RELIABILITY_CLAIM_GROUNDING_FLAGGED,
        metric.below_threshold,
    );
    span.record(
        attrs::FERRUMDECK_RELIABILITY_CLAIM_GROUNDING_THRESHOLD,
        metric.threshold,
    );
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_output_has_no_claims_and_never_flags() {
        let m = ClaimGrounding::compute_from_texts("", &[], 0.7);
        assert_eq!(m.claims_total, 0);
        assert_eq!(m.rate, 1.0);
        assert!(m.is_empty());
        assert!(!m.below_threshold, "no claims must never flag");
    }

    #[test]
    fn fully_grounded_output_rate_is_one() {
        let sources = vec!["Paris is the capital of France.".to_string()];
        let m =
            ClaimGrounding::compute_from_texts("The capital of France is Paris.", &sources, 0.7);
        assert_eq!(m.claims_total, 1);
        assert_eq!(m.claims_grounded, 1);
        assert_eq!(m.rate, 1.0);
        assert!(!m.below_threshold);
    }

    #[test]
    fn unsupported_claim_is_not_grounded() {
        let sources = vec!["Paris is the capital of France.".to_string()];
        let m = ClaimGrounding::compute_from_texts(
            "Bananas are purple and fly to the distant moon.",
            &sources,
            0.7,
        );
        assert_eq!(m.claims_total, 1);
        assert_eq!(m.claims_grounded, 0);
        assert_eq!(m.rate, 0.0);
        assert!(m.below_threshold);
    }

    #[test]
    fn trivial_fragments_are_dropped() {
        // "Done" / "OK" have < MIN_CLAIM_TOKENS significant tokens.
        let m = ClaimGrounding::compute_from_texts("Done. OK!", &[], 0.7);
        assert_eq!(m.claims_total, 0);
    }

    #[test]
    fn threshold_uses_strict_less_than() {
        // rate exactly == threshold must not flag.
        let m = ClaimGrounding::compute(10, 7, 0.70);
        assert!((m.rate - 0.70).abs() < 1e-9);
        assert!(
            !m.below_threshold,
            "exact-threshold rate must not flag (strict <)"
        );
    }

    #[test]
    fn grounded_clamped_to_total() {
        let m = ClaimGrounding::compute(3, 9, 0.7);
        assert_eq!(m.claims_grounded, 3);
        assert_eq!(m.rate, 1.0);
    }

    #[test]
    fn golden_cross_plane_fixture() {
        // Pinned alongside the Python golden test
        // (`tests/fixtures/claim_grounding.golden.json`): 3 claims, 2 grounded.
        let output = "The capital of France is Paris. The Eiffel Tower is in Paris. \
                      Bananas are purple and fly to the distant moon.";
        let sources = vec![
            "Paris is the capital of France.".to_string(),
            "The Eiffel Tower stands in Paris, France.".to_string(),
        ];
        let m = ClaimGrounding::compute_from_texts(output, &sources, 0.70);
        assert_eq!(m.claims_total, 3);
        assert_eq!(m.claims_grounded, 2);
        assert!((m.rate - 2.0 / 3.0).abs() < 1e-9);
        assert!(m.below_threshold, "0.667 < 0.70 → flagged");
    }

    #[test]
    fn serialises_stable_keys() {
        let m = ClaimGrounding::compute(3, 2, 0.7);
        let json = serde_json::to_value(m).unwrap();
        assert_eq!(json["claims_total"], 3);
        assert_eq!(json["claims_grounded"], 2);
        assert!((json["rate"].as_f64().unwrap() - 2.0 / 3.0).abs() < 1e-9);
        assert_eq!(json["below_threshold"], true);
    }

    #[test]
    fn record_on_span_compiles() {
        let span = tracing::Span::none();
        record_on_span(&span, &ClaimGrounding::compute(3, 2, 0.7));
    }
}
